### PR TITLE
Move from setup.py to pyproject.toml, switch to using ruff and tox, bump python version and update pre-commit hooks

### PR DESCRIPTION
**Pull request by e-belfer**
# Overview

Addresses subset of #118.

What problem does this address?
We have a super old and stale repo!

What did you change in this PR?

- Move from `setup.py` to `pyproject.toml`
- Move from various linters to `ruff`
- Update a bunch of the pre-commit configs
- Copy over some updated template actions from the cheshire repo
- Update the version of Python used in this repo to 3.12
- Move to using `tox`

Out of scope:
- Fix the failing integration tests that _aren't_ due to package updates. We'll want to fix these when we redo the processing sequences for both metrics sources.

# Questions for the reviewer
@bendnorman There's a remaining issue here with gcloud default settings not getting passed to the integration tests as expected. I'm anticipating that we'll rework these integration tests along with the actual extraction workflows as we get deeper into #118 - if you see the obvious problem here and want me to address this in the PR, let me know. Otherwise I'm inclined to merge in this already large set of changes and take up this issue as we continue. Thoughts?

# Testing

How did you make sure this worked? How can a reviewer verify this?

# To-do list

```[tasklist]
- [x] Fix all of the problematic ruff tests
- [x] Only test CI on one version of Python
- [x] Update python to 3.12
- [x] Update packages
- [x] Set up tox
- [x] Debug gcloud authentication
- [x] Figure out --cov tox failure
- [ ] Fix gcloud project config failure
- [ ] Review the PR yourself and call out any questions or issues you have
```

### CODE REVIEW

File: .github/workflows/tox-pytest.yml [hunk 1]
```diff
@@ -5,45 +5,58 @@ on: [push, pull_request]
...
+            - conda-forge
+            - defaults
+            channel_priority: strict
+
+      - name: Log conda environnment information
         run: |
-          mamba info
-          mamba list
+          conda info
+          conda list
```
Comment (bendnorman) [inline .github/workflows/tox-pytest.yml:37]: Why did you switch from mamba to conda?

File: src/usage_metrics/ops/datasette.py [hunk 2]
```diff
@@ -42,11 +41,31 @@ def extract(context) -> pd.DataFrame:
     start_date = context.op_config["start_date"]
     end_date = context.op_config["end_date"]
 
-    raw_logs = pandas_gbq.read_gbq(
+    sql = (
         "SELECT * FROM `datasette_logs.run_googleapis_com_requests`"
-        f" WHERE DATE(timestamp) >= '{start_date}' AND DATE(timestamp) < '{end_date}'",
-        project_id=project_id,
-        credentials=credentials,
```
Comment (bendnorman) [inline src/usage_metrics/ops/datasette.py:49]: Just curious, was this giving you a security warning or failure?

**Review by bendnorman — APPROVED** [review]
TYSM! This looks good. I think it makes sense to just merge this in and not work about the test failures for now given we're going to change a lot of the code.

As for the GCP failure, I'm not sure why the project id can't be accessed. Everything the  "Print GCloud config" step printed out looks correct to me. `google.auth.defaults()` should be grabbing the project id of the WIP resource. Might be a question for @jdangerx.

**Comment by bendnorman** [comment]
I think good next steps would be to rip out the intake code and move to using assets instead of ops.

File: .github/workflows/tox-pytest.yml (hunk 1 above)
Comment (e-belfer) [reply .github/workflows/tox-pytest.yml:37]: I was mirroring what we do in PUDL and the pudl archiver repos.

File: src/usage_metrics/ops/datasette.py (hunk 2 above)
Comment (e-belfer) [reply src/usage_metrics/ops/datasette.py:49]: It was giving me a security warning about SQL injections, yes!

**Pull request merged**